controllers/component_build_controller.go

```diff
@@ -152,17 +160,66 @@ func (r *ComponentBuildReconciler) Reconcile(ctx context.Context, req ctrl.Reque
 
 	if val, ok := component.Annotations[gitops.PaCAnnotation]; ok && val == "1" {
 		// Use pipelines as code build
-		r.Log.Info("PaC enabled")
-		if err := r.EnsurePACRepository(ctx, component); err != nil {
+		r.Log.Info("Pipelines as Code enabled")
+
+		// Obtain Pipelines as Code callback URL
+		webhookTargetUrl, err := r.getPaCRoutePublicUrl(ctx)
+		if err != nil {
+			return ctrl.Result{}, err
+		}
+
+		gitProvider, err := gitops.GetGitProvider(component)
+		if err != nil {
+			r.Log.Error(err, "error detecting git provider")
+			// Do not reconcile, because configuration must be fixed before it is possible to proceed.
+			return ctrl.Result{}, nil
+		}
+
+		// Expect that the secret contains token for Pipelines as Code webhook configuration,
+		// but under <git-provider>.token field. For example: github.token
+		// Also it can contain github-private-key and github-application-id
+		// in case GitHub Application is used instead of webhook.
+		pacSecret := corev1.Secret{}
+		if err := r.NonCachingClient.Get(ctx, types.NamespacedName{Namespace: pipelinesAsCodeNamespace, Name: gitopsprepare.PipelinesAsCodeSecretName}, &pacSecret); err != nil {
```

Comment:
So this is a tricky item, using the non caching client
1) on the side arguing for it, to use the caching client here, you need to set up a watch for secrets in the all the namespaces build could occur in; such watches will create **A LOT** of events
2) on the side arguing against it, using the non caching client means we are hitting the api server on all builds that use pipeline as code; if that becomes the norm, that is **A LOT** of extra activity on the api server (fwiw, in openshift the various controllers in OCP by default are instructed to not use the non caching client because of this)

I'm conflicted, but I think with the absence of testing at scale, this is probably the right initial choice to make, because of the simplicity on our end.  Plus, I honestly do not have a sense of how much the non caching client is being used across app studio.  Perhaps this is a drop in the bucket, and we'll mitigate this back making sure our hosted service has given sufficient processing power to the API server to handle.

But I mention all this, even though I am not asking for a change, just so we document that we have at least considered the pro and cons here @mmorhun

Reply:
@gabemontero thank you for the detailed review.
Let me explain my decision to use non caching client here. The line above is needed to read configured credential for Pipelines as Code. We need them in order to know how (for app or webhook) we should configure component git repository and to do the configuration itself (we need credentials from the secret). As you probably noticed, the PaC secret is not in the component namespace, but in global Pipelines as Code one, so it is related to PaC configuration rather than a component one. Now about reading secrets. If we use caching client, it caches ALL secrets in the cluster which is a huge amount of data. We used to have OOM problems on staging because the operator just tried to read a config map (usually a cluster has a lot of config maps and secrets, in short we'd be in trouble because amount of the operator RAM consumption would depend on cluster size. For our staging it was several gigabytes where normal consumption is one-two hundred megabytes). To overcome such huge caches, we may (and do use for other secrets withing the component namespace) custom cache configuration. But the only way to customize cache is to use a label selector, which means all secrets that are read by the operator must have the specified label. We use such label in local namespace for component related secrets, but we just cannot require the label on the global PaC secret outside the component namespace (that would brought additional burden for users and error prone, but if we could require the label, it would allow us to use only caching client in the implementation above. But in my understanding, it is not fair to ask for a build-service related label in global PaC config namespace).
However, there is a good piece of news: the code above (that uses non caching client) is called only on the component onboarding and not used on PaC builds.
I hopes this clears things behind my decision. Anyway, I am open for any suggestions to improve the code.

In the future, we might have a common label for whole AppStudio and require the label on global PaC secret (as PaC is a part of AppStudio deployment). But this should be done on higher level (I'll rise this topic on one of the next architecture calls).

Reply:
OK cool @mmorhun you appear on the same page wrt the cons of using the caching client, and detailing your thought process to using the non caching client confirms what I thought that you did not pick that without the sufficient consideration.

And yes, the additional clarification that the piece of code in question is only for component onboarding and not a per build thing is very good news, and IMO confirms this approach is OK  / we are good here.

Leave this open so other see and understand, or mark resolved to shorten your PR discussion length, at your discretion @mmorhun  - thanks

Reply:
Thanks for bringing this discussion, this was a very informative read.